rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+### Configuration
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
+}
+```
+
+`Db#put_put_with_options` can optionally specify a timestamp and a row-level TTL at insertion time 
+by leveraging newly added fields in `WriteOptions`:
+
+```rust
+pub struct WriteOptions {
+    // ...
+   
+    /// the timestamp (in millis) at which this operation is considered to be executed - 
+    /// note that SlateDB does not automatically resolve out-of-order insertions
+    /// (if you need to retrieve the most recent semantic insertion, specify the
+    /// option on ReadOptions)
+    ///
+    /// default: uses the clock time configured in the DbOptions
+    pub timestamp_ms: Option<i64>,
+   
+    /// the time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+`Db#get_with_options` can optionally specify: a grace period with which to continue searching the
+LSM tree for a more recent insertion. This allows you to specify an upper bound for what you expect
+out-of-order insertions to be.
+
+```rust
+pub struct ReadOptions {
+    // ...
+
+    /// if set to a positive number, SlateDB will not return the first matching row for
+    /// each key lookup and instead will continue to traverse the LSM tree until the next
+    /// matching key is older than the found key's timestamp + retrieval_grace_ms. Logically,
+    /// this is an "upper bound" on the magnitude of out-of-order insertions.
+    ///
+    /// Example: if there are three insertions [KeyA@100, KeyA@50, KeyA@150] that are each
```

Comment:
Won't this lead to weird behavior as compaction runs? What happens to reads if KeyA@50 or KeyA@150 are compacted out between reads with the same key/retrieval grace? It seems like it's possible to go from get() returning KeyA@150 to get() returning KeyA@100. Are you thinking that this feature will require a compactor that also has some bounded window where it keeps all key values that are in the window?

Reply:
It is possible to have consecutive reads return differing results with this feature, anything other than a MAX value on the grace_ms will not guarantee consistency (you can think of it as an optimization against scanning everything) unless you issue snapshot reads.

In the example given, if KeyA@50 is compacted with KeyA@150 then any non-zero grace_ms would return KeyA@150. The situation where it might change behavior is if:

```
L0 -> KeyA@50
L1 -> KeyA@100
L2 -> KeyA@150
```

In this case, if you issued it with grace_ms<=50 the first query would retrieve KeyA@100. Then if there's a compaction that merges L1/L2 the next query would return KeyA@150.